pkg/noderesourcetopology/cache/docs/NUMA-aware-scheduler-side-reserve-plugin.md

```diff
@@ -0,0 +1,340 @@
+
+1.  When a node is running the expected set of pods perhaps with some
+    free resources, but the desired workload is running
+2.  When a node is **correctly (as reflected by NRT object data)** fully
+    allocated
+3.  When a node is **pessimistically (as computed by the reserve
+    cache)** fully allocated
+
+Because of the pessimistic allocation, the scheduler will mark a node as
+fully allocated before it is actually the case.
+
+From the scheduler perspective, the node will reach a steady state, and
+in due time the kubelet will reach the same state. At this point, the
+NRT objects will be stable, and they will report the same state.
+
+The node state can be encoded:
+
+1.  As a plain string of unordered namespace/name pairs. This is the
+    simplest representation, but it's likely to be unwieldy
+
+2.  As a fingerprint. This representation is much more compact, but
+    needs to be computed as cheaply as possible and using the same steps
+    on both sides (node and scheduler)
+
+    a.  We don't cover how to compute the fingerprint, intentionally.
+        This task will be deferred by a third-party package. The
+        fingerprint needs to be stable and computable on both sides
+        independently. The hashing does need to be cryptographically
+        secure, and collisions will only cause increased load without
+        other negative effects.
+
+To help the scheduler, all NRT objects shall have a node state
+fingerprint to record the node state they represent.
+
+The fingerprint will be initially set in an annotation. Once this
+approach is fully validated and mature enough, we will propose a new
+field in the NodeResourceTopology object.
+
+## Cache invalidation algorithm
+
+1.  Watch for NRT updates. **Do not automatically reflect these updates
+    into the scheduler-side cache**. This update step must be explicit
+    and happen on well-defined conditions.
+
+2.  At reserve stage, account for the pod resources on all the node NUMA
+    zones. This usually means overallocation. If the node cache was
+    clean, mark it as dirty; if it was already dirty, it is still dirty.
+
+3.  Check the **cache invalidation condition: **when the cache is dirty
+    and a node is filtered out at least N times in a row, compute its
+    node state fingerprint
+
+    a.  It's easy to track how many consecutive times a node is filtered
+        out, and we use this value to approximate "node fully
+        allocated". Actually checking a node is "fully allocated"
+        would require setting detailed load thresholds which are
+        impractical or too complex.
+
+4.  Compare the computed node start fingerprint with the node state
+    fingerprint of the last NRT object received
+
+    a.  If they don't match, exit this flow
+
+    b.  If they match, overwrite the node cache with the content of the
+        > NRT object. The node cache is now clean again
+
+![mismatch](images/reserve6-mismatch.png)
+
+![update](images/reserve7-update.png)
```

Comment:
We should update the Fingerprint for Node 2 to `Fingerprint: Y2` and update the scheduled pod details (same mistake as above).

Reply:
updated as well